docs/bundler/hmr.md

```diff
@@ -0,0 +1,266 @@
+{% /callout %}
+
+Accepts hot updates for this module or its dependencies.
+
+```ts#hmr-accept.ts
+// Accept self (simplest form)
+import.meta.hot.accept();
+// This logs: "TODO: implement ImportMetaHot.accept" and does nothing
+
+// Accept self with a callback to handle updates
+import.meta.hot.accept(newModule => {
+  // This callback will not be called in the current implementation
+  console.log("Updated module received", newModule);
+});
+
+// Accept updates to specific dependencies
+import.meta.hot.accept(["./dep1.js", "./dep2.js"], ([dep1, dep2]) => {
+  // This callback will not be called in the current implementation
+  console.log("Dependencies updated");
+});
+```
+
+In future releases, when a module is accepted, Bun will properly respect the acceptance boundaries, preventing propagation of updates beyond accepted modules and executing the provided callbacks. This will enable more granular control over how module updates are handled.
+
+### `hot.dispose()`
+
+{% callout type="success" %}
+✅ **Fully implemented** — `hot.dispose()` works as expected, allowing you to register callbacks that will be executed before the module is replaced.
+{% /callout %}
+
+Registers a callback that will be called when the module is about to be replaced.
+
+```ts#dispose.ts
+import.meta.hot.dispose(data => {
+  // Clean up resources or save state
+  data.state = {
+    /* state to preserve */
+  };
+
+  // Teardown side effects
+  myEventListener.disconnect();
+
+  // Close connections
+  myWebSocket.close();
+
+  // Clear timers
+  clearTimeout(myTimer);
+});
+```
+
+The `data` object is passed to the callback and can be used to store information that will be available via `hot.data` in the next instance of the module. This is particularly useful for preserving state across module updates.
+
+### `hot.data`
+
+{% callout type="success" %}
+✅ **Fully implemented** — `hot.data` works as expected, preserving data between module evaluations through the values set in `dispose` callbacks.
+{% /callout %}
+
+An object that persists between updates. Contains data from the previous module instance that was passed in the `dispose` handler.
+
+```ts#counter.ts
+// Simple counter with hot reload state persistence
+let count = 0;
+
+// Restore previous count when hot reloaded
+if (import.meta.hot) {
+  // Recover previous count from hot.data
+  if (import.meta.hot.data.count !== undefined) {
+    count = import.meta.hot.data.count;
+  }
+
+  // Save state before module is replaced
+  import.meta.hot.dispose(data => {
+    data.count = count;
+  });
+
+  import.meta.hot.accept();
+}
+
+export function increment() {
+  return ++count;
+}
+```
+
+Without this pattern, the counter would reset to 0 each time the module is hot-reloaded. With `hot.data`, the counter's value persists across updates.
+
+This is one of the fully implemented features of Bun's HMR system, allowing you to maintain state even as modules are replaced.
```

Comment:
this comment is redundant.

in addition, i think this example from the react template is worthwhile showcasing

```
import { createRoot } from "react-dom/client";
import { StrictMode } from "react";
import { App } from "./App";

const elem = document.getElementById("root")!;
const app = (
  <StrictMode>
    <App />
  </StrictMode>
);

if (import.meta.hot) {
  // With hot module reloading, `import.meta.hot.data` is persisted.
  const root = (import.meta.hot.data.root ??= createRoot(elem));
  root.render(app);
} else {
  // The hot module reloading API is not available in production.
  createRoot(elem).render(app);
}
```